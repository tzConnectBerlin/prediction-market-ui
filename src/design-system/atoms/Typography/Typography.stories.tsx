import * as React from 'react';
// also exported from '@storybook/react' if you can deal with breaking changes in 6.1
import { Story, Meta } from '@storybook/react/types-6-0';
import { Typography, TypographyProps } from './Typography';

export default {
  title: 'Atom/Typography',
  component: Typography,
  argTypes: {},
} as Meta;

const Template: Story<TypographyProps> = (args) => <Typography {...args} />;

export const Default = Template.bind({});
Default.args = {
  children: 'ABCDEF',
};

export const Truncated = Template.bind({});
Truncated.args = {
  truncate: true,
  children: 'Lorem ipsum dolor sit amet consectetur adipisicing elit.'.repeat(20),
  isTruncated: (val) => console.log(val),
};
