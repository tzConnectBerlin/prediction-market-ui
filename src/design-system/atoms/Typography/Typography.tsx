import React, { useEffect } from 'react';
import styled from '@emotion/styled';
import {
  Typography as MaterialTypography,
  TypographyProps as MaterialTypographyProps,
} from '@material-ui/core';

const checkIfTruncated = (element: HTMLSpanElement) => element.scrollHeight > element.clientHeight;
export interface TypographyProps extends Omit<MaterialTypographyProps, 'variant'> {
  size?:
    | 'body1'
    | 'body2'
    | 'button'
    | 'caption'
    | 'h1'
    | 'h2'
    | 'h3'
    | 'h4'
    | 'h5'
    | 'h6'
    | 'inherit'
    | 'overline'
    | 'subtitle1'
    | 'subtitle2'
    | string;
  margin?: string;
  component?: unknown; // TODO: extract from material-ui props
  truncate?: boolean;
  truncateAfter?: number;
  isTruncated?: (value: boolean) => void | Promise<void>;
}

interface IStyledTypography {
  fontSize?: string;
  margin?: string;
  lines?: number;
}

const StyledTypography = styled(MaterialTypography)<IStyledTypography>`
  font-size: ${({ fontSize }) => fontSize};
  margin: ${({ margin }) => margin};
  &.truncate {
    overflow: hidden;
    display: -webkit-box;
    -webkit-box-orient: vertical;
    -webkit-line-clamp: ${({ lines }) => lines};
  }
`;

export const Typography: React.FC<TypographyProps> = ({
  size,
  children,
  truncate = false,
  truncateAfter = 3,
  className,
  isTruncated,
  ...rest
}) => {
  const fontSize = size && (size.includes('em') || size.includes('px')) ? size : undefined;
  const variant = (size && !fontSize ? size : 'body1') as MaterialTypographyProps['variant'];
  const classes = truncate ? [className, 'truncate'] : [className];
  const ref = React.createRef<HTMLSpanElement>();
  useEffect(() => {
    // TODO: find a better way than set timeout
    setTimeout(() => {
      if (ref.current && isTruncated) {
        isTruncated(checkIfTruncated(ref.current));
      }
    }, 100);
  }, [ref, isTruncated]);
  return (
    <StyledTypography
      className={classes.join(' ')}
      fontSize={fontSize as any}
      variant={variant}
      lines={truncateAfter}
      ref={ref}
      {...rest}
    >
      {children}
    </StyledTypography>
  );
};
