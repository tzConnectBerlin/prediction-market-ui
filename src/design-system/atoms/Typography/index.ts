export type { TypographyProps } from './Typography';
export { Typography, Typography as default } from './Typography';
