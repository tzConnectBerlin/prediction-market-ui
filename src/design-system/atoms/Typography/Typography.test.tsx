import renderer from 'react-test-renderer';
import { render } from '@testing-library/react';
import { Typography } from './Typography';

describe('Snapshot testing Typography Component', () => {
  it('renders correctly with small text', () => {
    const component = renderer.create(<Typography>ABCD</Typography>).toJSON();
    expect(component).toMatchSnapshot();
  });

  it('renders correctly with small truncate true', () => {
    const component = renderer
      .create(
        <Typography truncate>
          {'Lorem ipsum dolor sit amet consectetur adipisicing elit.'.repeat(20)}
        </Typography>,
      )
      .toJSON();
    expect(component).toMatchSnapshot();
  });
});

describe('Element testing Typography Component', () => {
  it('render correctly with small texts', async () => {
    const { getByText } = render(<Typography>ABCD</Typography>);
    expect(getByText(/ABCD/i)).toBeInTheDocument();
  });

  it('render correctly with custom size', async () => {
    const { getByText } = render(<Typography size="10em">ABCD</Typography>);
    expect(getByText(/ABCD/i)).toBeInTheDocument();
  });

  it('render correctly and truncates', async () => {
    const text = 'Lorem ipsum dolor sit amet consectetur adipisicing elit.'.repeat(20);
    const truncateMethod = jest.fn();
    render(
      <Typography truncate isTruncated={truncateMethod}>
        {text}
      </Typography>,
    );
    await new Promise((r) => setTimeout(r, 100));
    expect(truncateMethod).toBeCalled();
  });
});
