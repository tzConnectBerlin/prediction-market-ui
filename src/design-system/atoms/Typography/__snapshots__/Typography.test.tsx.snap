// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot testing Typography Component renders correctly with small text 1`] = `
.emotion-0 {
  margin: 0;
  font-family: "Roboto","Helvetica","Arial",sans-serif;
  font-weight: 400;
  font-size: 1rem;
  line-height: 1.5;
  letter-spacing: 0.00938em;
}

.emotion-0.truncate {
  overflow: hidden;
  display: -webkit-box;
  -webkit-box-orient: vertical;
  -webkit-line-clamp: 4;
}

<p
  className="MuiTypography-root MuiTypography-body1  emotion-0"
  lines={4}
>
  ABCD
</p>
`;

exports[`Snapshot testing Typography Component renders correctly with small truncate true 1`] = `
.emotion-0 {
  margin: 0;
  font-family: "Roboto","Helvetica","Arial",sans-serif;
  font-weight: 400;
  font-size: 1rem;
  line-height: 1.5;
  letter-spacing: 0.00938em;
}

.emotion-0.truncate {
  overflow: hidden;
  display: -webkit-box;
  -webkit-box-orient: vertical;
  -webkit-line-clamp: 4;
}

<p
  className="MuiTypography-root MuiTypography-body1  truncate emotion-0"
  lines={4}
>
  Lorem ipsum dolor sit amet consectetur adipisicing elit.Lorem ipsum dolor sit amet consectetur adipisicing elit.Lorem ipsum dolor sit amet consectetur adipisicing elit.Lorem ipsum dolor sit amet consectetur adipisicing elit.Lorem ipsum dolor sit amet consectetur adipisicing elit.Lorem ipsum dolor sit amet consectetur adipisicing elit.Lorem ipsum dolor sit amet consectetur adipisicing elit.Lorem ipsum dolor sit amet consectetur adipisicing elit.Lorem ipsum dolor sit amet consectetur adipisicing elit.Lorem ipsum dolor sit amet consectetur adipisicing elit.Lorem ipsum dolor sit amet consectetur adipisicing elit.Lorem ipsum dolor sit amet consectetur adipisicing elit.Lorem ipsum dolor sit amet consectetur adipisicing elit.Lorem ipsum dolor sit amet consectetur adipisicing elit.Lorem ipsum dolor sit amet consectetur adipisicing elit.Lorem ipsum dolor sit amet consectetur adipisicing elit.Lorem ipsum dolor sit amet consectetur adipisicing elit.Lorem ipsum dolor sit amet consectetur adipisicing elit.Lorem ipsum dolor sit amet consectetur adipisicing elit.Lorem ipsum dolor sit amet consectetur adipisicing elit.
</p>
`;
